File: rust/src/695.max-area-of-island.rs
use std::collections::{HashSet, VecDeque};

use crate::Solution;

#[allow(dead_code)]
impl Solution {
    pub fn max_area_of_island(grid: Vec<Vec<i32>>) -> i32 {
        let height = grid.len();
        let width = grid[0].len();

        let mut ans = 0;
        let mut visited: HashSet<(usize, usize)> = HashSet::new();

        for i in 0..height {
            for j in 0..width {
                if grid[i][j] != 1 || visited.contains(&(i, j)) {
                    continue;
                }

                let mut area = 0;
                let mut iland = VecDeque::new();
                iland.push_back((i, j));

                while let Some((y, x)) = iland.pop_front() {
                    if visited.contains(&(y, x)) {
                        continue;
                    }

                    area += 1;
                    visited.insert((y, x));

                    let directions = [(0, 1), (1, 0), (!0, 0), (0, !0)];
                    for &(dy, dx) in directions.iter() {
                        let x = x.wrapping_add(dx);
                        let y = y.wrapping_add(dy);

                        if x < width && y < height && grid[y][x] == 1 {
                            iland.push_back((y, x));
                        }
                    }
                }

                ans = ans.max(area);
            }
        }
        ans
    }
}

File: ruby/695.max-area-of-island.rb
# frozen_string_literal: true

DIRECTIONS = [[0, 1], [1, 0], [-1, 0], [0, -1]].freeze

# @param {Integer[][]} grid
# @return {Integer}
def max_area_of_island(grid)
  height = grid.size
  width = grid[0].size
  ans = 0
  visited = Set.new

  grid.each_with_index do |row, i|
    row.each_with_index do |cell, j|
      next if cell.zero? || visited.include?([i, j])

      area = 0
      iland = [[i, j]]

      until iland.empty?
        y, x = iland.shift
        next if visited.include?([y, x])

        area += 1
        visited.add([y, x])

        DIRECTIONS.each do |dy, dx|
          new_x = x + dx
          new_y = y + dy

          next if new_x.negative? || new_x >= width
          next if new_y.negative? || new_y >= height
          next if grid[new_y][new_x] != 1

          iland.push([new_y, new_x])
        end

      end
      ans = [ans, area].max
    end
  end

  ans
end
